### PR TITLE
Java: Add FieldValueNode to break up cartesian step relation.

### PR DESCRIPTION
**Pull request by aschackmull**
This improves performance when there are fields that have both a high number of reads and a high number of assignments.

### CODE REVIEW

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll [hunk 1]
```diff
@@ -44,7 +44,8 @@ newtype TNode =
   } or
   TSummaryInternalNode(SummarizedCallable c, FlowSummaryImpl::Private::SummaryNodeState state) {
     FlowSummaryImpl::Private::summaryNodeRange(c, state)
-  }
+  } or
+  TFieldValueNode(Field f)
```
Comment (hvitved) [inline java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll:48]: Restrict to static fields?

Comment (hvitved) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll:48]: I guess that doesn't work as you also use it in e.g. `DispatchFlow`.

Comment (aschackmull) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowNodes.qll:48]: Exactly.

File: java/ql/lib/semmle/code/java/dataflow/internal/DataFlowPrivate.qll [hunk 2]
```diff
@@ -205,7 +211,30 @@ class CastNode extends ExprNode {
   CastNode() { this.getExpr() instanceof CastExpr }
 }
 
-class DataFlowCallable = Callable;
+private newtype TDataFlowCallable =
+  TCallable(Callable c) or
+  TFieldScope(Field f)
```
[the remark this answers is not in the document]
Comment (aschackmull) [reply java/ql/lib/semmle/code/java/dataflow/internal/DataFlowPrivate.qll:216]: Yes, but I'm not sure we can guarantee the existence of one. And even if we can, we'd have to jump through some hoops to select a unique arbitrary constructor.

**Review by hvitved — APPROVED** [review]

**Pull request merged**